File: src/main/java/org/cubexmc/metro/lifecycle/MapIntegrationLifecycle.java
package org.cubexmc.metro.lifecycle;

import java.util.List;
import java.util.Locale;

import org.cubexmc.metro.Metro;
import org.cubexmc.metro.integration.BlueMapIntegration;
import org.cubexmc.metro.integration.DynmapIntegration;
import org.cubexmc.metro.integration.MapIntegration;
import org.cubexmc.metro.integration.SquaremapIntegration;
import org.cubexmc.metro.util.SchedulerUtil;

/**
 * Owns optional web map integrations and their refresh lifecycle.
 */
public class MapIntegrationLifecycle {

    private static final String AUTO_PROVIDER = "AUTO";
    private static final List<String> AUTO_PROVIDER_ORDER = List.of("BLUEMAP", "DYNMAP", "SQUAREMAP");

    private final Metro plugin;
    private final IntegrationFactory integrationFactory;
    private final RefreshScheduler refreshScheduler;
    private MapIntegration activeIntegration;
    private String activeProvider;
    private boolean refreshQueued;
    private Object refreshTaskId;

    public MapIntegrationLifecycle(Metro plugin) {
        this(plugin, provider -> createDefaultIntegration(plugin, provider), new SchedulerUtilRefreshScheduler());
    }

    MapIntegrationLifecycle(Metro plugin, IntegrationFactory integrationFactory, RefreshScheduler refreshScheduler) {
        this.plugin = plugin;
        this.integrationFactory = integrationFactory;
        this.refreshScheduler = refreshScheduler;
    }

    public void enable() {
        activateConfiguredProvider();
    }

    public void disable() {
        cancelQueuedRefresh();
        if (activeIntegration != null) {
            activeIntegration.disable();
        }
        activeIntegration = null;
        activeProvider = null;
    }

    public void refresh() {
        try {
            if (!activateConfiguredProvider()) {
                return;
            }
            activeIntegration.refresh();
        } catch (Throwable e) {
            plugin.getLogger().warning("[Map] Failed to refresh " + activeProvider
                    + " integration: " + e.getMessage());
        }
    }

    public void requestRefresh() {
        if (plugin.getConfigFacade() == null || !plugin.getConfigFacade().isMapIntegrationEnabled() || refreshQueued) {
            return;
        }
        refreshQueued = true;
        long delay = plugin.getConfigFacade().getMapRefreshDelayTicks();
        refreshTaskId = refreshScheduler.schedule(plugin, () -> {
            refreshQueued = false;
            refreshTaskId = null;
            refresh();
        }, delay, -1L);
    }

    private void cancelQueuedRefresh() {
        if (refreshTaskId != null) {
            refreshScheduler.cancel(refreshTaskId);
            refreshTaskId = null;
        }
        refreshQueued = false;
    }

    private boolean activateConfiguredProvider() {
        if (plugin.getConfigFacade() == null || !plugin.getConfigFacade().isMapIntegrationEnabled()) {
            disable();
            return false;
        }

        String provider = normalizeProvider(plugin.getConfigFacade().getMapProvider());
        if (provider == null || provider.isBlank()) {
            plugin.getLogger().warning("[Map] map_integration.provider is empty. Skipping map integration.");
            disable();
            return false;
        }

        if (AUTO_PROVIDER.equals(provider)) {
            return activateAutoProvider();
        }

        if (activeIntegration != null && provider.equals(activeProvider)) {
            return true;
        }

        disable();
        try {
            activeIntegration = integrationFactory.create(provider);
            if (activeIntegration == null) {
                plugin.getLogger().warning("[Map] Unknown map provider '" + provider
                        + "'. Expected AUTO, BLUEMAP, DYNMAP, or SQUAREMAP.");
                return false;
            }
            if (!activeIntegration.isAvailable()) {
                plugin.getLogger().info("[Map] " + provider + " is not available, skipping map integration.");
                activeIntegration = null;
                return false;
            }
            activeProvider = provider;
            activeIntegration.enable();
            return true;
        } catch (Throwable e) {
            plugin.getLogger().info("[Map] " + provider + " API not found, skipping map integration.");
            activeIntegration = null;
            activeProvider = null;
            return false;
        }
    }

    private boolean activateAutoProvider() {
        if (activeIntegration != null && activeProvider != null && activeIntegration.isAvailable()) {
            return true;
        }

        disable();
        for (String provider : AUTO_PROVIDER_ORDER) {
            try {
                MapIntegration integration = integrationFactory.create(provider);
                if (integration == null || !integration.isAvailable()) {
                    continue;
                }
                activeIntegration = integration;
                activeProvider = provider;
                plugin.getLogger().info("[Map] AUTO selected " + provider + " map provider.");
                activeIntegration.enable();
                return true;
            } catch (Throwable e) {
                plugin.getLogger().info("[Map] " + provider + " API not found, skipping map integration.");
            }
        }
        plugin.getLogger().warning("[Map] AUTO provider could not find BlueMap, Dynmap, or Squaremap.");
        return false;
    }

    private String normalizeProvider(String provider) {
        return provider == null ? null : provider.trim().toUpperCase(Locale.ROOT);
    }

    private static MapIntegration createDefaultIntegration(Metro plugin, String provider) {
        return switch (provider) {
            case "BLUEMAP" -> new BlueMapIntegration(plugin);
            case "DYNMAP" -> new DynmapIntegration(plugin);
            case "SQUAREMAP" -> new SquaremapIntegration(plugin);
            default -> null;
        };
    }

    @FunctionalInterface
    interface IntegrationFactory {
        MapIntegration create(String provider);
    }

    interface RefreshScheduler {
        Object schedule(Metro plugin, Runnable task, long delay, long period);

        void cancel(Object taskId);
    }

    private static final class SchedulerUtilRefreshScheduler implements RefreshScheduler {
        @Override
        public Object schedule(Metro plugin, Runnable task, long delay, long period) {
            return SchedulerUtil.globalRun(plugin, task, delay, period);
        }

        @Override
        public void cancel(Object taskId) {
            SchedulerUtil.cancelTask(taskId);
        }
    }
}

File: src/main/java/org/cubexmc/metro/lifecycle/MapIntegrationLifecycle.kt
package org.cubexmc.metro.lifecycle

import java.util.Locale
import org.cubexmc.metro.Metro
import org.cubexmc.metro.integration.BlueMapIntegration
import org.cubexmc.metro.integration.DynmapIntegration
import org.cubexmc.metro.integration.MapIntegration
import org.cubexmc.metro.integration.SquaremapIntegration
import org.cubexmc.metro.util.SchedulerUtil

/**
 * Owns optional web map integrations and their refresh lifecycle.
 */
class MapIntegrationLifecycle {
    private val plugin: Metro
    private val integrationFactory: IntegrationFactory
    private val refreshScheduler: RefreshScheduler
    private var activeIntegration: MapIntegration? = null
    private var activeProvider: String? = null
    private var refreshQueued = false
    private var refreshTaskId: Any? = null

    constructor(plugin: Metro) : this(
        plugin,
        IntegrationFactory { provider -> createDefaultIntegration(plugin, provider) },
        SchedulerUtilRefreshScheduler(),
    )

    constructor(
        plugin: Metro,
        integrationFactory: IntegrationFactory,
        refreshScheduler: RefreshScheduler,
    ) {
        this.plugin = plugin
        this.integrationFactory = integrationFactory
        this.refreshScheduler = refreshScheduler
    }

    fun enable() {
        activateConfiguredProvider()
    }

    fun disable() {
        cancelQueuedRefresh()
        activeIntegration?.disable()
        activeIntegration = null
        activeProvider = null
    }

    fun refresh() {
        try {
            if (!activateConfiguredProvider()) {
                return
            }
            activeIntegration?.refresh()
        } catch (e: Throwable) {
            plugin.logger.warning("[Map] Failed to refresh $activeProvider integration: ${e.message}")
        }
    }

    fun requestRefresh() {
        if (plugin.configFacade == null || !plugin.configFacade.isMapIntegrationEnabled() || refreshQueued) {
            return
        }
        refreshQueued = true
        val delay = plugin.configFacade.getMapRefreshDelayTicks()
        refreshTaskId = refreshScheduler.schedule(
            plugin,
            Runnable {
                refreshQueued = false
                refreshTaskId = null
                refresh()
            },
            delay,
            -1L,
        )
    }

    private fun cancelQueuedRefresh() {
        val taskId = refreshTaskId
        if (taskId != null) {
            refreshScheduler.cancel(taskId)
            refreshTaskId = null
        }
        refreshQueued = false
    }

    private fun activateConfiguredProvider(): Boolean {
        if (plugin.configFacade == null || !plugin.configFacade.isMapIntegrationEnabled()) {
            disable()
            return false
        }

        val provider = normalizeProvider(plugin.configFacade.getMapProvider())
        if (provider == null || provider.isBlank()) {
            plugin.logger.warning("[Map] map_integration.provider is empty. Skipping map integration.")
            disable()
            return false
        }

        if (AUTO_PROVIDER == provider) {
            return activateAutoProvider()
        }

        if (activeIntegration != null && provider == activeProvider) {
            return true
        }

        disable()
        try {
            val integration = integrationFactory.create(provider)
            activeIntegration = integration
            if (integration == null) {
                plugin.logger.warning("[Map] Unknown map provider '$provider'. Expected AUTO, BLUEMAP, DYNMAP, or SQUAREMAP.")
                return false
            }
            if (!integration.isAvailable()) {
                plugin.logger.info("[Map] $provider is not available, skipping map integration.")
                activeIntegration = null
                return false
            }
            activeProvider = provider
            integration.enable()
            return true
        } catch (e: Throwable) {
            plugin.logger.info("[Map] $provider API not found, skipping map integration.")
            activeIntegration = null
            activeProvider = null
            return false
        }
    }

    private fun activateAutoProvider(): Boolean {
        val currentIntegration = activeIntegration
        if (currentIntegration != null && activeProvider != null && currentIntegration.isAvailable()) {
            return true
        }

        disable()
        for (provider in AUTO_PROVIDER_ORDER) {
            try {
                val integration = integrationFactory.create(provider)
                if (integration == null || !integration.isAvailable()) {
                    continue
                }
                activeIntegration = integration
                activeProvider = provider
                plugin.logger.info("[Map] AUTO selected $provider map provider.")
                integration.enable()
                return true
            } catch (e: Throwable) {
                plugin.logger.info("[Map] $provider API not found, skipping map integration.")
            }
        }
        plugin.logger.warning("[Map] AUTO provider could not find BlueMap, Dynmap, or Squaremap.")
        return false
    }

    private fun normalizeProvider(provider: String?): String? = provider?.trim()?.uppercase(Locale.ROOT)

    fun interface IntegrationFactory {
        fun create(provider: String): MapIntegration?
    }

    interface RefreshScheduler {
        fun schedule(plugin: Metro, task: Runnable, delay: Long, period: Long): Any?

        fun cancel(taskId: Any)
    }

    private class SchedulerUtilRefreshScheduler : RefreshScheduler {
        override fun schedule(plugin: Metro, task: Runnable, delay: Long, period: Long): Any? =
            SchedulerUtil.globalRun(plugin, task, delay, period)

        override fun cancel(taskId: Any) {
            SchedulerUtil.cancelTask(taskId)
        }
    }

    companion object {
        private const val AUTO_PROVIDER = "AUTO"
        private val AUTO_PROVIDER_ORDER = listOf("BLUEMAP", "DYNMAP", "SQUAREMAP")

        private fun createDefaultIntegration(plugin: Metro, provider: String): MapIntegration? =
            when (provider) {
                "BLUEMAP" -> BlueMapIntegration(plugin)
                "DYNMAP" -> DynmapIntegration(plugin)
                "SQUAREMAP" -> SquaremapIntegration(plugin)
                else -> null
            }
    }
}
